docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+
+There's nothing special about the method `newbord` in comparison with others, apart from the fact that it allocates, initialises and returns an object of the same type.
+
+Destructors are still under debate, but they won't have to deallocate memory like in C++, just execute finalization tasks like closing file handles, running some cleanup routines, checking final consistency, etc.
+
+## Primitive Types
+
+Verona treats numeric types specially in the compiler but their representaiton isn't special in the language.
+
+Verona's numeric types are singleton types (no fields) that have their machine representaiton based on their names.
+So, `I32` is a signed integer with 32-bit width (in LLVM's parlour, an `i32`) and so on.
+
+The full list of primitive numeric types are:
+ * Unsigned integers: `U8`, `U16`, `U32`, `U64` and `U128`
+ * Signed integers: `I8`, `I16`, `I32`, `I64` and `I128`
+ * Floating Point: `F32` and `F64`
+ * System types: `ULength` and `ILength` (machine-word sized integers)
+
+Note, integer literals have no defined type and their types (and representation) will be inferred by the compiler depending on the type of the variable being assigned to.
+
+## Type Unions
+
+Type unions means a variable can be either type on the list.
+
+For example:
+```ts
+foobar(a : (A | B | C))
+{
+  ...
+}
+
+main()
+{
+  let a : A;
+  let b : B;
+  let d : D;
+
+  // This is fine
+  foobar(a);
+  foobar(b);
+
+  // This is not
+  foobar(d); // ERROR
+}
+```
+
+The imlementation of `foobar` will have to cope with all three types, either via a `match` statement or taking advantage of some behaviour that is common to all three (say, an interface they all implement).
+
+The compiler is free, however, to optimise the calls by creating two versions of `foobar` and replace the calls to the generic `foobar` for their specific ones.
+If the semantics is defined at compile time, this can simplifies a `match` statement into their specific matches and even inline the call altogether.
+
+Example:
+```ts
+// The generic one, if there are still dynamic cases
+foobar(a : (A | B | C)) { ... }
+
+// The specific specialisations used in defined cases
+foobar(a : A) { ... }
+foobar(b : B) { ... }
+```
+
+## Type Intersections
+
+Intersection means a type must implement all guarantees of all types involved.
+
+For example:
+```ts
+interface Serialisable { toString() : String; };
+
+main()
+{
+  let x : (A & Serialisable & mut) = ...;
+}
+```
+
+In the above example, the type of `x` must conform to both `A` and `Serialisable` guarantees (methods, fields) as well as the `mutable` capability.
+
+Every type in Verona must have a capability, so:
+```ts
+var x : (A & imm); // this is ok
+
+var y : B; // this is not
+```
+
+Type alias can be used to make it easier to use types, for example:
+```ts
+class _String { ... };
+
+// We mostly want strings as values
+type String = (_String & imm);
+
+// But we may want to change it
+type MutableString = (_String & mut);
+```
```

Comment:
This seems unrelated to explaining type intersections.

Reply:
They're not _explaining_ intersections, they're just demonstrating how you can use aliases to not have to declare capabilities on types that are commonly associated with one capability and not others.